File: problems-labs/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long cardnum = get_long("Number: ");
    long divisor = 10, adder = 0, bigsum = 0, smallsum = 0;
    const long cardnumcopy = cardnum;
    int digits = 0;
    string cardtype;
//Get number of digits in card number
    while (cardnum / divisor > 0)
    {
        cardnum /= divisor;
        digits++;
    }
    digits++;
//Determine card type
    if (digits == 15)
    {
        if (cardnumcopy / 10000000000000 == 34 || cardnumcopy / 10000000000000 == 37)
        {
            cardtype = "AMEX\n";
        }
        else
        {
            cardtype = "INVALID\n";
        }
    }
    else if (digits == 16)
    {
        if (cardnumcopy / 100000000000000 == 51 || cardnumcopy / 100000000000000 == 52 || cardnumcopy / 100000000000000 == 53
            || cardnumcopy / 100000000000000 == 54 || cardnumcopy / 100000000000000 == 55)
        {
            cardtype = "MASTERCARD\n";
        }
        else if (cardnum == 4)
        {
            cardtype = "VISA\n";
        }
            else
        {
            cardtype = "INVALID\n";
        }
    }
    else if (digits == 13)
    {
        if (cardnum == 4)
        {
            cardtype = "VISA\n";
        }
    }
    else
    {
        cardtype = "INVALID\n";
    }

    cardnum = cardnumcopy;  // gets original card number

    while (cardnum / divisor > 0)
    {
        cardnum /= divisor;
        adder = ((cardnum % 10) * 2);
        if (adder / 10 != 0)
        {
            adder = ((adder % 10) + (adder / 10));
        }
        bigsum += adder;
        divisor = 100;
    }
    divisor = 1;
    cardnum = cardnumcopy;

    while (cardnum / divisor > 0)
    {
        cardnum /= divisor;
        adder = (cardnum % 10);
        smallsum += adder;
        divisor = 100;
    }

    int checksum = bigsum + smallsum;
    if (checksum % 10 == 0)
    {

    }
    else
    {
        cardtype = "INVALID\n";
    }
    printf("%s", cardtype);
}

File: problems-labs/sentimental-credit/credit.py
# TODO
# Problem: https://cs50.harvard.edu/x/2022/psets/6/credit/
import re
from cs50 import get_string


def main():
    # define acceptable card prefixes
    # MASTERCARD
    patternm = "(51|52|53|54|55)"
    # AMEX
    patterna = "(34|37)"
    # VISA
    patternv = "4"

    number = get_string("Number: ")
    cardnum = int(number)

    # adder = 0
    # bigsum = 0
    # smallsum = 0

    # check if number matches a regular expression
    if (re.search(patternm, number) and len(number) == 16):
        type = "MASTERCARD"
    elif (re.search(patterna, number) and len(number) == 15):
        type = "AMEX"
    elif (re.search(patternv, number) and (len(number) == 13 or len(number) == 16)):
        type = "VISA"
    else:
        type = "INVALID"

    divisor = 1

    # compute sums for Luhn's algorithm
    checksum = sum1(cardnum) + sum2(cardnum)

    # if managed to pass regex check but invalidated by Luhn's algorithm
    if (checksum % 10 != 0):
        type = "INVALID"

    print(type)


def sum1(cardnum):
    divisor = 10
    bigsum = 0
    while (cardnum / divisor > 0):
        cardnum = int(cardnum / divisor)
        adder = ((cardnum % 10) * 2)
        if int(adder / 10) != 0:
            adder = (adder % 10) + (int(adder / 10))
        bigsum += adder
        divisor = 100
    return bigsum


def sum2(cardnum):
    divisor = 1
    smallsum = 0
    while int(cardnum / divisor) > 0:
        cardnum = int(cardnum / divisor)
        adder = cardnum % 10
        smallsum += adder
        divisor = 100
    return smallsum


main()
